src/control/server/storage/bdev/runner.go

```diff
@@ -99,36 +116,39 @@ func defaultScriptRunner(log logging.Logger) *spdkSetupScript {
 //
 // NOTE: will make the controller disappear from /dev until reset() called.
 func (s *spdkSetupScript) Prepare(req *storage.BdevPrepareRequest) error {
-	nrHugepages := req.HugePageCount
-	// Always supply non-zero number of hugepages in request otherwise devices cannot be
-	// accessed.
-	if nrHugepages <= 0 {
-		nrHugepages = defaultNrHugepages
+	s.env = map[string]string{
+		"PATH":          os.Getenv("PATH"),
+		pciBlockListEnv: req.PCIBlockList,
 	}
 
-	env := []string{
-		fmt.Sprintf("PATH=%s", os.Getenv("PATH")),
-		fmt.Sprintf("%s=%d", nrHugepagesEnv, nrHugepages),
-		fmt.Sprintf("%s=%s", targetUserEnv, req.TargetUser),
-	}
-	if req.HugeNodes != "" {
-		env = append(env, fmt.Sprintf("%s=%s", hugeNodeEnv, req.HugeNodes))
-	}
-	if req.PCIAllowList != "" {
-		env = append(env, fmt.Sprintf("%s=%s", pciAllowListEnv, req.PCIAllowList))
-	}
-	if req.PCIBlockList != "" {
-		env = append(env, fmt.Sprintf("%s=%s", pciBlockListEnv, req.PCIBlockList))
-	}
 	if req.DisableVFIO {
-		env = append(env, fmt.Sprintf("%s=%s", driverOverrideEnv, vfioDisabledDriver))
+		s.env[driverOverrideEnv] = vfioDisabledDriver
+	} else if req.EnableVMD {
+		// Run setup with DRIVER_OVERRIDE=none to speed up VMD re-binding as per
+		// https://github.com/spdk/spdk/commit/b0aba3fcd5aceceea530a702922153bc75664978.
+		s.env[driverOverrideEnv] = noDriver
+
+		// Apply block list to cater for situation where VMD is configured for use with
+		// DAOS but other NVMe drives should be reserved for other use (bdev_exclude).
+		if err := s.run(); err != nil {
+			return errors.Wrap(err, "vmd driver reset")
+		}
+
+		delete(s.env, driverOverrideEnv)
 	}
 
-	s.log.Debugf("spdk setup env: %v", env)
-	out, err := s.runCmd(s.log, env, s.scriptPath)
-	s.log.Debugf("spdk setup stdout:\n%s\n", out)
+	s.env[targetUserEnv] = req.TargetUser
+	s.env[pciAllowListEnv] = req.PCIAllowList
 
-	return errors.Wrapf(err, "spdk setup failed (%s)", out)
+	// Always use min number of hugepages otherwise devices cannot be accessed.
+	nrHugepages := req.HugePageCount
+	if nrHugepages <= 0 {
+		nrHugepages = defaultNrHugepages
+	}
+	s.env[nrHugepagesEnv] = fmt.Sprintf("%d", nrHugepages)
+	s.env[hugeNodeEnv] = req.HugeNodes
```

Comment:
Just wondering why we no longer set these at the top? Can't use them for the VMD setup call?

Reply:
The initial VMD call just unbinds the backing devices from the kernel, and does not rebind. Therefore no other parameters are required. We do want to supply the block list in the initial VMD prep call to cater for the case where someone wants to keep an NVMe device bound to the kernel for use via the OS outside of DAOS.

Reply:
splitting up the methods made this more obvious